source/server/hot_restart_impl.cc

```diff
@@ -138,319 +137,35 @@ HotRestartImpl::HotRestartImpl(const Options& options)
   }
   stats_allocator_ = std::make_unique<Stats::RawStatDataAllocator>(stat_lock_, *stats_set_,
                                                                    options_.statsOptions());
-  my_domain_socket_ = bindDomainSocket(options.restartEpoch());
-  child_address_ = createDomainSocketAddress((options.restartEpoch() + 1));
-  initDomainSocketAddress(&parent_address_);
-  if (options.restartEpoch() != 0) {
-    parent_address_ = createDomainSocketAddress((options.restartEpoch() + -1));
-  }
-
   // If our parent ever goes away just terminate us so that we don't have to rely on ops/launching
   // logic killing the entire process tree. We should never exist without our parent.
   int rc = prctl(PR_SET_PDEATHSIG, SIGTERM);
   RELEASE_ASSERT(rc != -1, "");
 }
 
-int HotRestartImpl::bindDomainSocket(uint64_t id) {
-  Api::OsSysCalls& os_sys_calls = Api::OsSysCallsSingleton::get();
-  // This actually creates the socket and binds it. We use the socket in datagram mode so we can
-  // easily read single messages.
-  int fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK, 0);
-  sockaddr_un address = createDomainSocketAddress(id);
-  Api::SysCallIntResult result =
-      os_sys_calls.bind(fd, reinterpret_cast<sockaddr*>(&address), sizeof(address));
-  if (result.rc_ != 0) {
-    throw EnvoyException(
-        fmt::format("unable to bind domain socket with id={} (see --base-id option)", id));
-  }
-
-  return fd;
-}
-
-void HotRestartImpl::initDomainSocketAddress(sockaddr_un* address) {
-  memset(address, 0, sizeof(*address));
-  address->sun_family = AF_UNIX;
-}
-
-sockaddr_un HotRestartImpl::createDomainSocketAddress(uint64_t id) {
-  // Right now we only allow a maximum of 3 concurrent envoy processes to be running. When the third
-  // starts up it will kill the oldest parent.
-  const uint64_t MAX_CONCURRENT_PROCESSES = 3;
-  id = id % MAX_CONCURRENT_PROCESSES;
-
-  // This creates an anonymous domain socket name (where the first byte of the name of \0).
-  sockaddr_un address;
-  initDomainSocketAddress(&address);
-  StringUtil::strlcpy(&address.sun_path[1],
-                      fmt::format("envoy_domain_socket_{}", options_.baseId() + id).c_str(),
-                      sizeof(address.sun_path) - 1);
-  address.sun_path[0] = 0;
-  return address;
-}
-
 void HotRestartImpl::drainParentListeners() {
-  if (options_.restartEpoch() > 0) {
-    // No reply expected.
-    RpcBase rpc(RpcMessageType::DrainListenersRequest);
-    sendMessage(parent_address_, rpc);
-  }
-
+  as_child_.drainParentListeners();
```

Comment:
Other than changing the version constant, the changes here in hot_restart_impl.cc are entirely just pulling code out into the new classes HotRestarting{Base,Child,Parent}. You'll see that most of the public functions have been replaced with passthroughs to either HotRestartingChild as_child_ or HotRestartingParent as_parent_, as appropriate.